File: FunkoShop_Exam/src/test/java/es/codeurjc/test/exam/FunkoShop/FunkoShopTest.java
package es.codeurjc.test.exam.FunkoShop;

import static org.junit.jupiter.api.Assertions.*;
import static org.mockito.Mockito.*;

import java.util.Optional;

import org.junit.jupiter.api.*;

public class FunkoShopTest {

    private FunkoRepository funkoRepository;
    private FunkoShop funkoShop;

    @BeforeEach
    public void setUp() {
        funkoRepository = mock(FunkoRepository.class);
        funkoShop = new FunkoShop(funkoRepository);
    }

    @Test
    public void testBuyFunkoSuccessfully() {
        // Arrange
        Funko funko = new Funko("Batman", "15.99", true);
        when(funkoRepository.findByName("Batman")).thenReturn(Optional.of(funko));

        // Act
        Funko purchasedFunko = funkoShop.buy("Batman");

        // Assert
        assertNotNull(purchasedFunko);
        assertEquals("Batman", purchasedFunko.getName());
        assertFalse(purchasedFunko.isAvailable());
        verify(funkoRepository).save(funko);
    }

    @Test
    public void testBuyFunkoNotAvailable() {
        // Arrange
        Funko funko = new Funko("Superman", "12.99", false);
        when(funkoRepository.findByName("Superman")).thenReturn(Optional.of(funko));

        // Act & Assert
        RuntimeException exception = assertThrows(RuntimeException.class, () -> funkoShop.buy("Superman"));
        assertEquals("Funko not available", exception.getMessage());
        verify(funkoRepository, never()).save(any());
    }

    @Test
    public void testBuyFunkoNotFound() {
        // Arrange
        when(funkoRepository.findByName("Wonder Woman")).thenReturn(Optional.empty());

        // Act & Assert
        RuntimeException exception = assertThrows(RuntimeException.class, () -> funkoShop.buy("Wonder Woman"));
        assertEquals("Funko not found", exception.getMessage());
        verify(funkoRepository, never()).save(any());
    }
}

File: test_integracion/FunkoShop_Exam/src/test/FunkoShop.test.js
const { FunkoShop } = require('../src/FunkoShop'); // Ajusta la ruta según tu proyecto
const { FunkoRepository } = require('../src/FunkoRepository'); // Ajusta la ruta según tu proyecto
const { Funko } = require('../src/Funko'); // Ajusta la ruta según tu proyecto

jest.mock('../src/FunkoRepository'); // Mock del repositorio

describe('FunkoShop', () => {
  let funkoRepository;
  let funkoShop;

  beforeEach(() => {
    funkoRepository = new FunkoRepository();
    funkoShop = new FunkoShop(funkoRepository);
  });

  test('should buy a Funko successfully', () => {
    // Arrange
    const funko = new Funko('Batman', '15.99', true);
    funkoRepository.findByName = jest.fn().mockReturnValue(Promise.resolve(funko));
    funkoRepository.save = jest.fn();

    // Act
    return funkoShop.buy('Batman').then((purchasedFunko) => {
      // Assert
      expect(purchasedFunko).not.toBeNull();
      expect(purchasedFunko.name).toBe('Batman');
      expect(purchasedFunko.isAvailable).toBe(false);
      expect(funkoRepository.save).toHaveBeenCalledWith(funko);
    });
  });

  test('should throw an error if the Funko is not available', () => {
    // Arrange
    const funko = new Funko('Superman', '12.99', false);
    funkoRepository.findByName = jest.fn().mockReturnValue(Promise.resolve(funko));

    // Act & Assert
    return expect(funkoShop.buy('Superman')).rejects.toThrow('Funko not available');
  });

  test('should throw an error if the Funko is not found', () => {
    // Arrange
    funkoRepository.findByName = jest.fn().mockReturnValue(Promise.resolve(null));

    // Act & Assert
    return expect(funkoShop.buy('Wonder Woman')).rejects.toThrow('Funko not found');
  });
});